src/sqlfluff/rules/L050.py

```diff
@@ -0,0 +1,115 @@
+"""Implementation of Rule L050."""
+from typing import Optional
+
+from sqlfluff.core.rules.base import BaseRule, LintResult, LintFix, RuleContext
+from sqlfluff.core.rules.doc_decorators import document_fix_compatible
+
+
+@document_fix_compatible
+class Rule_L050(BaseRule):
+    """Files must not begin with newlines or whitespace.
+
+    | **Anti-pattern**
+    | The content in file begins with newlines or whitespace, the ^ represents the beginning of file.
+
+    .. code-block:: sql
+       :force:
+
+        ^
+
+        SELECT
+            a
+        FROM foo
+
+        -- Beginning on an indented line is also forbidden,
+        -- (the • represents space).
+
+        ••••SELECT
+        ••••a
+        FROM
+        ••••foo
+
+    | **Best practice**
+    | Start file on either code or comment, the ^ represents the beginning of file.
+
+    .. code-block:: sql
+       :force:
+
+
+        ^SELECT
+            a
+        FROM foo
+
+        -- Including an initial block comment.
+
+        ^/*
+        This is a description of my SQL code.
+        */
+        SELECT
+            a
+        FROM
+            foo
+
+        -- Including an initial inline comment.
+
+        ^--This is a description of my SQL code.
+        SELECT
+            a
+        FROM
+            foo
+    """
+
+    @staticmethod
+    def _potential_template_collision(context: RuleContext) -> bool:
+        """Check for any templated raw slices that intersect with source slices in the raw_stack.
+
+        Returns:
+            :obj:`bool` indicating a preceding templated raw slice has been detected.
+
+        """
+        templated_file = context.segment.pos_marker.templated_file
+        for segment in context.raw_stack:
+            source_slice = segment.pos_marker.source_slice
+            raw_slices = templated_file.raw_slices_spanning_source_slice(source_slice)
+            if any(
+                raw_slice
+                for raw_slice in raw_slices
+                if raw_slice.slice_type == "templated"
+            ):
```

Comment:
I've a few questions on this code (my Python is pretty poor so apologies if these are all non-issues):
1. Should the first `raw_slice` be `raw_slices`? If not how can you reference `raw_slice` before it's defined? Or is the point, it's false initially but `any()` will pick up one of the other ones? And then it might be true later?
2. There's nothing under the `for` loop? Or is that taken care of by the `any()`?
3. With the use of `any()` doe that mean the last line is not a necessity for this? Does that mean a non-templated piece could be seen as templated?

Sorry just finding it hard to grok this code. If this is correct, and common enough in Python, then happy to leave as is. Though might get @barrywhart to approve this PR then as a bit out of my comfort zone.

Reply:
So this is a pretty common use of a generator to find whether any element in an iterable satisfies a condition, it's very similar to list comprehension.

It means: return each `raw_slice` in the list of `raw_slices` only if that `raw_slice` satisfies the condition `raw_slice.slice_type == "templated"`. The `any` means that as soon as one raw_slice is returned from the generator then the condition is satisfied so no need to check the rest of `raw_slices` and enter the `if` block

Reply:
Cool. My lack of Python experience here showing. Resolving!